### PR TITLE
Add GlobalValueNumbering pass

### PR DESCRIPTION
**Pull request by hkapp**
Add the whole implementation for GVN, comprised of the definition of idempotent (might be a problem, please look at the comment), deep equality of variables, and the pass itself.

The Debug part that are useful should be put somewhere else, hints on "where" would be appreciated.

### CODE REVIEW

**Comment by densh** [comment]
You forgot to mention that this fixes #32.

File: tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala [hunk 1]
```diff
@@ -120,4 +393,91 @@ object GlobalValueNumbering {
 
   }
 
+  object Debug {
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala]: Those should go to `Shows`.

**Comment by hkapp** [comment]
I can't put the Debug methods in nir.Shows, as nir does not depend on tools (and it really shouldn't). Any ideas ?

**Comment by densh** [comment]
@hkapp Lets put them in `compiler.analysis.Shows` then.

**Comment by hkapp** [comment]
@densh Ok, fixed it. What would you say about the comment I added in `isIdempotent` ? I really think that this invalidates the definition of idempotency for `call`, and that we won't be able to reuse these values until there is static analysis available for this specific purpose.

File: tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala [hunk 2]
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
...
+      val cfg        = ControlFlow.Graph(defn.insts)
+      val domination = DominatorTree.build(cfg)
+
+      val newInsts = performSimpleValueNumbering(cfg, domination)
+
+      Seq(defn.copy(insts = newInsts))
+  }
+
+  private def pred(block: Block) = block.pred.map(_.from)
+  private def succ(block: Block) = block.succ.map(_.to)
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala]: Lets rename `block.pred` into `block.inEdges` and `block.succ` into `block.outEdges` and then put `def pred` and `def succ` as helper methods directly inside `ControlFlow.Block`.

File: tools/src/main/scala/scala/scalanative/compiler/analysis/Shows.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+object Shows {
+
+  private def blockToString(block: Block): String =
+    showLocal(block.name).toString
+
+  def showCFG(cfg: ControlFlow.Graph): String = {
+    cfg.all.map { block =>
+      val succStr =
+        block.succ.map(_.to).map(blockToString).mkString("(", ",", ")")
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/analysis/Shows.scala]: `block.succ.map(_.to)` can be simplified to `block.pred` after refactoring I've suggested earlier.

File: tools/src/main/scala/scala/scalanative/compiler/analysis/Shows.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      val predStr =
+        block.pred.map(_.from).map(blockToString).mkString("(", ",", ")")
+      s"${blockToString(block)} -> ${succStr}, pred = ${predStr}"
+    }.mkString("\n")
+  }
+
+  def showDominatorTree(domination: Map[Block, Set[Block]]): String = {
+    domination.toList
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/analysis/Shows.scala]: `toSeq`

File: tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala [hunk 5]
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
...
+  def apply(ctx: Ctx) = new GlobalValueNumbering()(ctx.top)
+
+  /* The definition of idempotent for Call might be wrong.
+   * Example:
+   * val a = new AnyRef { var b: Int = 0; override def toString = b.toString }
+   * val c = a.toString
+   * a.b = 1
+   * val d = a.toString
+   * --> toString is pure, but the values of c and d are different !
+   * */
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala]: `toString` is not pure, `load`-s on external memory are side-effecting by definition.

File: tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala [hunk 6]
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
...
+        true
+
+      // Never idempotent:
+      case (_: Load | _: Store | _: Stackalloc | _: Classalloc) => false
+
+      // Special cases:
+      case Call(_, Val.Global(Ref(node), _), _) =>
+        node.attrs.isPure
+
+      case Call(_, _, _) => false
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala]: `case _: Call => false`

File: tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala [hunk 7]
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
...
+
+      case _: Closure => ???
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala]: Lets assume that this one is not idempotent for now.

File: tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala [hunk 8]
```diff
@@ -2,11 +2,283 @@ package scala.scalanative
...
+      if (!(isIdempotent(opA) && isIdempotent(opB)))
+        false
+      else {
+        (opA, opB) match {
+          // Here we know that the function called is idempotent
+          case (Call(tyA, ptrA, argsA), Call(tyB, ptrB, argsB)) =>
+            eqType(tyA, tyB) && eqVal(ptrA, ptrB) && eqVals(argsA, argsB)
+
+          case (Load(_, _), Load(_, _)) =>
+            false // not idempotent
```
Comment (densh) [inline tools/src/main/scala/scala/scalanative/compiler/pass/GlobalValueNumbering.scala]: It seems that this and all the other not-idempotent cases are never going to happen due to the check above, therefore we should drop this code altogether.

**Comment by hkapp** [comment]
Looking at the code in ControlFlow.Block, there are still a few variables named `node(s)` instead of `block`. I also saw that most of them have been changed, but in `Graph.apply` they are both used to denote different things. Any semantic meaning to this ?

**Comment by densh** [comment]
We used to have `ControFlow.Node`, that has been subsumed by `ControlFlow.Block`. Node is a block, and visa versa.

**Comment by densh** [comment]
LGTM, great job @hkapp !

**Pull request merged**